agent/service.go

```diff
@@ -84,6 +98,50 @@ func (ts *TaskService) Create(ctx context.Context, req *shimapi.CreateTaskReques
 	return resp, nil
 }
 
+func (ts *TaskService) proxyStdio(ctx context.Context, stdin, stdout, stderr string) {
+	go proxyIO(ctx, stdin, internal.StdinPort, true)
+	go proxyIO(ctx, stdout, internal.StdoutPort, false)
+	go proxyIO(ctx, stderr, internal.StderrPort, false)
+}
+
+func proxyIO(ctx context.Context, path string, port uint32, in bool) {
+	if path != "" {
+		log.G(ctx).Debug("setting up IO for " + path)
+		f, err := fifo.OpenFifo(ctx, path, syscall.O_RDWR|syscall.O_NONBLOCK|syscall.O_CREAT, 0700)
+		if err != nil {
+			log.G(ctx).WithError(err).Error("error opening fifo")
+			return
+		}
+		listener, err := vsock.Listen(port)
+		if err != nil {
+			log.G(ctx).WithError(err).Error("unable to listen on vsock")
+			f.Close()
+			return
+		}
+		conn, err := listener.Accept()
+		if err != nil {
+			listener.Close()
+			f.Close()
+			log.G(ctx).WithError(err).Error("error accepting vsock connection")
+		}
+		go func() {
+			<-ctx.Done()
+			conn.Close()
+			f.Close()
```

Comment:
How do you cancel this `ctx` ?

Reply:
It's not cancelled atm. I'll update it to cancel before it calls TaskService.Cancel() in Shutdown. Does that seem like the right place?

Reply:
This has already been rebased on top of the commit and has that cancel. I don't think it helps though since that cancel is passed in and not generated/controlled by the TaskService itself.

Would it make sense to, instead of having NewTaskService() take a cancelFunc, to return one to the caller so that the TaskService maintains control of what the cancel does?

Reply:
We need same `context` to be passed to `server.Serve(ctx, listener)` and `server.Shutdown(ctx)` in order to stop properly everything.

How about something like this:
```go
taskService, shutdownSignal := NewTaskService(runcTaskService) <- no cancel here

group.Go(func() error {
     return server.Serve(ctx, listener)
})

...

group.Go(func() error {
     <- shutdownSignal
     cancel() <--  call cancel here instead of calling it from `Shutdown`
})
```

Reply:
I did something a little different by holding a slice of []context.CancelFuncs and calling them all during shutdown, adding the one that is passed in by NewTaskService() to the list. Does that seem like it will work?